src/proxy/inbound.rs

```diff
@@ -155,91 +156,158 @@ impl Inbound {
+            debug!("connected to: {}", ri.upstream_addr);
+            Ok((conn_guard, stream))
+        };
+        let (conn_guard, mut stream) = match rx.await {
+            Ok(res) => res,
+            Err(InboundFlagError(err, flag, code)) => {
+                ri.result_tracker.record_with_flag(Err(err), flag);
+                if let Err(err) = req.send_error(build_response(code)) {
+                    tracing::warn!("failed to send HTTP response: {err}");
+                }
+                return;
+            }
         };
 
-        let destination_workload = pi.local_workload_information.get_workload().await?;
+        // At this point, we established the upstream connection and need to send a 200 back to the client.
+        // we may still have failures at this point during the proxying, but we don't need to send these
+        // at the HTTP layer.
+        let send = req
+            .send_response(build_response(StatusCode::OK))
+            .and_then(|h2_stream| async {
+                if ri.inbound_protocol == AppProtocol::PROXY {
+                    let Connection {
+                        src, src_identity, ..
+                    } = ri.rbac_ctx.conn;
+                    super::write_proxy_protocol(&mut stream, (src, ri.hbone_addr), src_identity)
+                        .instrument(trace_span!("proxy protocol"))
+                        .await?;
+                }
+                copy::copy_bidirectional(
+                    h2_stream,
+                    copy::TcpStreamSplitter(stream),
+                    &ri.result_tracker,
+                )
+                .instrument(trace_span!("hbone server"))
+                .await
+            });
+        let res = conn_guard.handle_connection(send).await;
+        ri.result_tracker.record(res);
+    }
 
-        if let Err(e) =
-            Self::validate_destination(&pi.state, &conn, &destination_workload, hbone_addr).await
-        {
-            metrics::log_early_deny(conn.src, conn.dst, Reporter::destination, e);
-            return req.send_error(build_response(StatusCode::BAD_REQUEST));
+    async fn build_inbound_request(
+        pi: &Arc<ProxyInputs>,
+        conn: Connection,
+        req: &H2Request,
+    ) -> Result<InboundRequest, InboundError> {
+        if req.method() != Method::CONNECT {
+            let e = Error::NonConnectMethod(req.method().to_string());
+            return Err(InboundError(e, StatusCode::BAD_REQUEST));
         }
 
+        let start = Instant::now();
+        let hbone_addr = req
+            .uri()
+            .to_string()
+            .as_str()
+            .parse::<SocketAddr>()
+            .map_err(|_| {
+                InboundError(
+                    Error::ConnectAddress(req.uri().to_string()),
+                    StatusCode::BAD_REQUEST,
+                )
+            })?;
+
+        let destination_workload = pi
+            .local_workload_information
+            .get_workload()
+            .await
```

Comment:
Would this be infallible? It seems like we shouldn't get to a point where the ztunnel is accepting connections of behalf of a workload it doesn't know about.

Reply:
get_workload does a join with XDS, so that may not have landed yet

Reply:
Added a comment about it, but i don't think its a safe enough assumption to `expect()` it

Reply:
There could be future TLS providers that don't depend on the local workload, etc